jscomps/PrivateDataManager.js

```diff
@@ -30,6 +34,63 @@ PrivateDataManager.prototype = {
     return Cc["@mozilla.org/cookiemanager;1"].getService(Ci.nsICookieManager);
   },
 
+  _cacheSizeObserver: null,
+
+  _cacheSizePromise: null,
+
+  get cacheSizePromise () {
+    if (this._cacheSizePromise) {
+      return this._cacheSizePromise;
+    }
+
+    this._cacheSizePromise = new Promise((resolve, reject) => {
+      // Needs to root the observer since cache service keeps only a weak reference.
+      this._cacheSizeObserver = {
+        onNetworkCacheDiskConsumption: consumption => {
+          resolve(consumption);
+          this._cacheSizePromise = null;
+          this._cacheSizeObserver = null;
+        },
+
+        QueryInterface: ChromeUtils.generateQI([
+          Ci.nsICacheStorageConsumptionObserver,
+          Ci.nsISupportsWeakReference
+        ])
+      };
+
+      try {
+        Services.cache2.asyncGetDiskConsumption(this._cacheSizeObserver);
+      } catch (e) {
+        reject(e);
+        this._cacheSizePromise = null;
+        this._cacheSizeObserver = null;
+      }
+    });
+
+    return this._cacheSizePromise;
+  },
+
+  get siteDataSizePromise() {
+    var _siteDataSizePromise = new Promise((resolve, reject) => {
+      try {
+        Services.qms.getUsage(function (request) {
+          let usage = 0;
+          if (request.resultCode == Cr.NS_OK) {
+            let items = request.result;
+            for (let item of items) {
+              usage += item.usage;
+            }
+          }
+          resolve(usage);
```

Comment:
Make this orthogonal with the cacheSizePromise when it comes to nullifying the promise.

I'd add "_siteDataSizePromise: null," instead of local variable and 

```
    if (this._siteDataSizePromise) {
      return this._siteDataSizePromise;
    }
   
    this._siteDataSizePromise = new Promise((resolve, reject) => {
      ...
      resolve(usage);
      this._siteDataSizePromise = null;
      ...
      reject(e);
      this._siteDataSizePromise = null;
      ...
    }

```